### PR TITLE
Taproot wallet test vectors (generation+tests)

### PR DESCRIPTION
**Pull request by sipa**
This PR adds code to `test/functional/feature_taproot.py` which runs through a (deterministic) scenario covering several aspects of the wallet side of BIP341 (scriptPubKey computation from keys/scripts, control block computation, key path spending), with the ability to output test vectors in mediawiki format based on this scenario. The generated tests are then also included directly in `src/test/script_tests.cpp` and `src/test/script_standard_tests.cpp`.

The test vectors generated here were added to BIP341 in https://github.com/bitcoin/bips/pull/1225

### CODE REVIEW

**Comment by sipa** [comment]
Inspired by @giacomocaironi's https://github.com/bitcoin/bips/pull/1191.

File: test/functional/test_framework/key.py [hunk 1]
```diff
@@ -368,15 +379,18 @@ def get_pubkey(self):
         ret.compressed = self.compressed
         return ret
 
-    def sign_ecdsa(self, msg, low_s=True):
+    def sign_ecdsa(self, msg, low_s=True, rfc6979=False):
```
Comment (junderw) [inline test/functional/test_framework/key.py:382]: Why was this added? It isn't used anywhere in the changes.

Comment (sipa) [reply test/functional/test_framework/key.py:382]: It is, indirectly through the "deterministic" flag in the feature_taproot's context, which on its turn is used in the test vector scenario (to make sure the fully signed transaction is exactly the same every run).

Comment (junderw) [reply test/functional/test_framework/key.py:382]: My mistake. I see it now. Thanks.

**Comment by sipa** [comment]
I'm going to convert this to construct tests in JSON format rather than Mediawiki, and also actually run it in that form directly.

File: test/functional/feature_taproot.py [hunk 2]
```diff
@@ -1616,41 +1606,40 @@ def gen_test_vectors(self):
...
+            given['internalPubkey'] = tap.internal_pubkey.hex()
+            def pr(node):
+                if node is None:
+                    return None
+                elif isinstance(node, tuple):
+                    return {"id": int(node[0]), "script": node[1].hex(), "leafVersion": node[2]}
+                elif len(node) == 1:
+                    return pr(node[0])
+                elif len(node) == 2:
+                    return [pr(node[0]), pr(node[1])]
```
Comment (junderw) [inline test/functional/feature_taproot.py:1623]: Mentioned in the other PR.

```suggestion
            def pr(node):
                if node is None:
                    return []
                elif isinstance(node, tuple):
                    return {"id": int(node[0]), "script": node[1].hex(), "leafVersion": node[2]}
                elif len(node) == 1:
                    return [pr(node[0])]
                elif len(node) == 2:
                    return [pr(node[0]), pr(node[1])]
```

File: test/functional/feature_taproot.py [hunk 3]
```diff
@@ -1481,10 +1495,237 @@ def test_spenders(self, node, spenders, input_counts):
...
+        # Generate corresponding public x-only pubkeys
+        pubs = [compute_xonly_pubkey(prv)[0] for prv in prvs]
+        # Generate taproot objects
+        inner_keys = [pubs[i] for i in range(7)]
+        script_lists = [
+            None,
+            [("0", CScript([pubs[50], OP_CHECKSIG]), 0xc0)],
+            [("0", CScript([pubs[51], OP_CHECKSIG]), 0xc0)],
+            [("0", CScript([pubs[52], OP_CHECKSIG]), 0xc0), ("1", CScript([b"BIP341"]), pubs[99][0] & 0xfe)],
+            [("0", CScript([pubs[53], OP_CHECKSIG]), 0xc0), ("1", CScript([b"Taproot"]), pubs[99][1] & 0xfe)],
```
Comment (sanket1729) [inline test/functional/feature_taproot.py:1532]: It just happens that pubs[99][0] &0xfe and pubs[99][1] & 0xfe are not colliding with annex tag (0x50). Maybe add an assert along with the other asserts that the above is not the case?

Comment (sipa) [reply test/functional/feature_taproot.py]: Fixed, by restricting the selection to the set of leaf versions recommended in the BIP.

**Comment by sipa** [comment]
Updated to address https://github.com/bitcoin/bips/pull/1225#pullrequestreview-799478503

File: src/key.cpp [hunk 4]
```diff
@@ -263,6 +263,7 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
 
 bool CKey::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint256* merkle_root, const uint256* aux) const
```
Comment (laanwj) [inline src/key.cpp:264]: Instead of changing the meaning of `aux==nullptr` here, what about:

- Remove the default arguments from `CKey::SignSchnorr` (they're unused)
- Disallow nullptr by changing the argument to a `const uint256& aux`
- Pass zero in from the caller if it wants to use zero

This seems more explicit and clear to me? (The behavior is not documented in the doc comment, in any case, right now)

Comment (sipa) [reply src/key.cpp]: Done.

**Comment by sipa** [comment]
Rebased.

**Comment by sipa** [comment]
Updated to address https://github.com/bitcoin/bips/pull/1225#discussion_r747460936.

**Comment by sipa** [comment]
Corresponding BIP change was merged: https://github.com/bitcoin/bips/pull/1225

**Comment by laanwj** [comment]
Code review ACK f1c33ee4ac1056289f2e67b75755388549ada4ca
Also checked that `src/test/data/bip341_wallet_vectors.json` matches the file in bip341.

**Comment by laanwj** [comment]
Closing: PR was merged, github didn't detect it
